form-controls-acc-name: Modernise the rule (#1357)
act-rules/act-rules.github.io: _rules/form-control-non-empty-accessible-name-e086e5.md

> This rule applies to any element that is [included in the accessibility tree](#included-in-the-accessibility-tree), and that has one of the following [semantic roles][]: `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch`, `textbox`.

Reviewer: I question why menuitemcheckbox and menuitemradio appear here and menuitem does not.

Author: Fair question. Similarly I think it's worth asking why this include spinbutton but not button. This is unrelated to the PR though, so I'll open a separate issue for it. Definitely something worth looking at.

- Certain assistive technologies can be set up to ignore the title attribute, which means that to some users the title attribute will not act as an [accessible name][].
- Several assistive technologies have a functionality to list all form fields on a page, including the `disabled` ones. Therefore this rule is still applicable to `disabled` form fields. If an assistive technology consistently ignores `disabled` form fields in all its interactions, then it is possible to have a `disabled` form field with no accessible name without creating accessibility issues for the user.
- Implementation of [Presentational Roles Conflict Resolution][] varies from one browser or assistive technology to another. Depending on this, some elements can have one of the applicable [semantic roles][] and fail this rule with some technology but users of other technologies would not experience any accessibility issue.

> - Element with the `option` role are not tested in this rule because options do not meet the definition of a [User interface component](https://www.w3.org/TR/WCAG21/#dfn-user-interface-components). If these elements are presented as user interface components, these need to be tested separate from this rule.

Reviewer 2: ```suggestion
- Elements with the `option` role are not tested in this rule because they do not meet the definition of a [User interface component](https://www.w3.org/TR/WCAG21/#dfn-user-interface-components). If these elements are presented as user interface components, these need to be tested separately from this rule.
```

## Background

The list of roles in the applicability is derived by taking all the roles from [WAI-ARIA Specifications](#wai-aria-specifications) that:

> - have a [semantic roles][] that inherits from the [abstract](https://www.w3.org/TR/wai-aria/#abstract_roles) `input` or `select` role, and

Reviewer 2: ```suggestion
- have [semantic roles][] that inherit from the [abstract](https://www.w3.org/TR/wai-aria/#abstract_roles) `input` or `select` role, and
```

> - does not have a [required context](https://www.w3.org/TR/wai-aria/#scope) role that itself inherits from one of those roles.

Reviewer 2: ```suggestion
- do not have a [required context](https://www.w3.org/TR/wai-aria/#scope) role that itself inherits from one of those roles.
```

> - have [semantic roles][] that inherit from the [abstract](https://www.w3.org/TR/wai-aria/#abstract_roles) `input` or `select` role, and

Reviewer: https://github.com/w3c/aria/pull/1278 removes the multiple inheritance from menuitemcheckbox and menuitemradio such that this will no longer be true once this appears in a spec.

Author: Thanks for the heads up! I'll open an issue for it. This'll be actionable once ARIA 1.2 makes it further into rec track.

This rule does not map to [3.3.2 Labels or Instructions](https://www.w3.org/TR/WCAG21/#labels-or-instructions) as there are sufficient techniques within 3.3.2 that don't need the elements to have an [accessible name][]. For example "G131: Providing descriptive labels" **AND** "G162: Positioning labels to maximize predictability of relationships" would be sufficient.

- [Understanding Success Criterion 4.1.2: Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value)
- [H91: Using HTML form controls and links](https://www.w3.org/WAI/WCAG21/Techniques/html/H91)
- [H44: Using label elements to associate text labels with form controls](https://www.w3.org/WAI/WCAG21/Techniques/html/H44)
- [H65: Using the title attribute to identify form controls when the label element cannot be used](https://www.w3.org/WAI/WCAG21/Techniques/html/H65)

## Test Cases

### Passed

#### Passed Example 1

> This `input` element has an [accessible name][] because of its implicit label.

Reviewer 3: I'd say programmatic is more accurate. Implicit could be a suggested visual label which is not programmatically associated.

#### Passed Example 5

This element with a `combobox` [role][semantic role] has an [accessible name][] because of its `aria-label` attribute.

```html
<div>Country</div>
<div aria-label="country" role="combobox" aria-disabled="true">England</div>
```

### Failed

#### Failed Example 1

> This `input` element has no attribute to give it an [accessible name][].

Reviewer 4: ```suggestion
This `input` element does not have an attribute that gives an [accessible name][] to it.
```

```html
<div>last name</div>
<input />
```

#### Failed Example 2

> This disabled `input` element has no attribute to give it an [accessible name][].

Reviewer 4: ```suggestion
This disabled `input` element does not have an attribute that gives an [accessible name][] to it.
```

```html
<input disabled />
```

#### Failed Example 3

> This element with a `combobox` [role][semantic role] has an empty (`""`) [accessible name][] because the space in the `aria-label` [attribute value][] is trimmed.

Reviewer 2: There is no element with a combobox role in this example

#### Passed Example 5

> This element with a `combobox` [role][semantic role] has an [accessible name][] because of its `aria-label` attribute.

Reviewer 3: Isn't the first role link redundant?

Author: Why would it be?

Reviewer 3: Ignore my comment, I didn't realise that "[role]" was a link to semantic role definition... sorry!

```html
<div>Country</div>
<div aria-label="country" role="combobox" aria-disabled="true">England</div>
```

### Failed

#### Failed Example 1

This `input` element has no attribute to give it an [accessible name][].

```html
<div>last name</div>
<input />
```

#### Failed Example 2

> This [non-focusable](#focusable) `input` element has no attribute to give it an [accessible name][].

Reviewer 5: "non-fousable" is not really correct. It's out of sequential focus navigation, but still programmatically focusable.

Author: I'm replacing this with a disabled element. Makes for a better example IMO.

#### Failed Example 6

This element with a `textbox` [role][semantic role] has an empty (`""`) [accessible name][]. The `label` element does not give it an [accessible name][], this only works for native form controls.

```html
<label for="lastname">first name</label>
<div role="textbox" id="lastname"></div>
```

### Inapplicable

#### Inapplicable Example 1

> This `input` element is not [included in the accessibility tree][] because of its `style` attribute.

Reviewer 4: It's not actually because of `style` attribute alone but because of `display: none;`
```suggestion
This `input` element is not [included in the accessibility tree][] because of its `style` attribute with `display:none;`.
```

Reviewer 3: Or... "display property value set to none".
